### PR TITLE
Pint convert input

### PR DESCRIPTION
**Pull request by rcaneill**
WIP PR
Add a check / conversion that inputs have the proper unit

### CODE REVIEW

**Comment by rcaneill** [comment]
So my idea was:
1. have a dict with input units
2. verify that units are correct (only for quantities), and if necessary convert them

And then when it works, in another PR parse the matlab doc to get all input units

**Comment by DocOtak** [comment]
With the caveat that I don't have a lot of pint experience in practice...

Do you know how "safe" it is to automatically convert units? GSW is pretty specific with what the inputs of functions should be, I feel that a standard_name + unit would give enough information to "know" what to do, but units alone aren't enough. I suspect that if, for example, your absolute salinities inputs are in mg/kg there is something funny going on and we should probably throw an exception saying that the units are wrong.

**Comment by rcaneill** [comment]
> Do you know how "safe" it is to automatically convert units? GSW is pretty specific with what the inputs of functions should be, I feel that a standard_name + unit would give enough information to "know" what to do, but units alone aren't enough. I suspect that if, for example, your absolute salinities inputs are in mg/kg there is something funny going on and we should probably throw an exception saying that the units are wrong.

I see your point, and indeed for SA the unit of input datasets should probably always be g/kg. But for e.g. pressure multiple units are commonly used, and it would make sense to me to convert them.

**Comment by rcaneill** [comment]
I haven't checked the consistency (yet, but I can do it).

**Comment by rcaneill** [comment]
To summarize #48 
(cf https://github.com/DocOtak/gsw-xarray/issues/48#issuecomment-1142541752)
> OK, to summarize:
> 
> * If any input is a Quantity, all inputs must be Quantities
>       
>   * Quantities will be converted silently as if we had wrapped things in `@ureg.wraps`
> 
> * If no inputs are Quantities, assume input values are correct for what GSW wants (basically what we do now)
> 
>  * If you don't want Unit conversion, "dequantify" your function inputs.

**Comment by DocOtak** [comment]
If the units are consistent with variable names, let's give option 2 a try.

Also, do you think it would make architectural sense if we had a unit mapping factory function when we know that pint is being used rather than a dict of strs that is converted later?

Something like the following:
```python
def gsw_units(ureg):
   # ureg is the unit registry that is being used by the inputs to some gsw function
   return {
    "CT": ureg.Unit("degC"),
    "SA": ureg.Unit("g/kg"),
    "SP": ureg.Unit("1")  # or ureg.Unit("dimensionless")
   }
```

**Comment by DocOtak** [comment]
Yeah... tired for me too.

We should probably just use the `Quantity.to()` method and pass it a string of the correct units.

**Comment by rcaneill** [comment]
(TODO list for me)
* [ ] Check that all args are pint if one of them is pint
* [ ] Add unit for all arguments of all functions
* [ ] Add test to verify that all functions are unit mapped
* [ ] Add doc

**Comment by rcaneill** [comment]
I am just missing the doc part for now (I may have time tomorrow).
@DocOtak if you can review the code this would be great. I think that I did everything on the todo list and according to discussion #48, but I may have missed something.

**Comment by rcaneill** [comment]
I think that the only thing left is how to deal with the units `degree_east` and `degree_north`. If the user uses cf_xarray registry it will not be a problem, but how to deal with the other case?

**Comment by DocOtak** [comment]
Just wanted to let you know that I saw this and can hopefully get to it this weekend.... just got off a ship and readjusting to shore.

File: gsw_xarray/_core.py [hunk 1]
```diff
@@ -36,55 +38,93 @@ def quantify(rv, attrs, unit_registry=None):
...
     return new_args, new_kwargs, registries
 
 
-def cf_attrs(attrs, name, check_func):
+def cf_attrs(fname, attrs, name, check_func):
     def cf_attrs_decorator(func):
         @wraps(func)
         def cf_attrs_wrapper(*args, **kwargs):
-            args, kwargs, unit_registry = pint_compat(args, kwargs)
+            args_names = get_args_names(func, args)
```
Comment (DocOtak) [inline gsw_xarray/_core.py:126]: It looks like machinery for converting everything into kwargs was being made. (the args_and_kwargs_to_args func in _function_utils.py). Should we maybe convert all inputs into a dict of kwargs and just work with that internally? I might be unaware of some edge cases, but it feels like we could get rigd of a lot of duplication in the helper functions that handle the args and kwargs seperately.

File: gsw_xarray/_core.py [hunk 2]
```diff
@@ -36,55 +38,93 @@ def quantify(rv, attrs, unit_registry=None):
     return rv
 
 
-def pint_compat(args, kwargs):
+def dequantify_reg(kw, arg):
```
Comment (DocOtak) [inline gsw_xarray/_core.py:41]: This almost feels like a good candidate for functools.singledispatch, though that uses the type of the first argument

File: gsw_xarray/_core.py (hunk 1 above)
Comment (rcaneill) [reply gsw_xarray/_core.py:126]: Yes you are right, I had the same idea, and then wanted to keep the args just to stay the more general as possible. But I don't think that any of the gsw functions use args that can't be transformed into kwargs (I mean there is no gsw function defined by: `def _(arg1, arg2, *, kwargs1=None):`).

I will convert every arg to kwarg

File: gsw_xarray/_core.py (hunk 2 above)
Comment (rcaneill) [reply gsw_xarray/_core.py:41]: I may be possible to use singledispatch but I am not sure that we would gain a lot here.

I had another idea (but not very precise) that I wanted to (try to) implement after this PR: refactor the pint_compat, add_attr, quantify, etc into decorators, so the code would look like:
```python
@pint_compat
@dequantify
def funcwrap(func, args, kwargs):
    # call func(args, kwargs)
    pass
```
I guess we have to split out functions into 2 groups: functions working before calling gsw.function - dequantify, pint_compat - and fucntions working after - quantify, add_attrs -

File: gsw_xarray/_core.py (hunk 1 above)
Comment (rcaneill) [reply gsw_xarray/_core.py]: done

File: gsw_xarray/_core.py (hunk 2 above)
Comment (rcaneill) [reply gsw_xarray/_core.py:41]: Sorry I think that I missed your point... I thought that you were proposing to use functools.singledispatch for the wrapper.
It seems that it is well appropriated for `dequantify_reg`

Comment (rcaneill) [reply gsw_xarray/_core.py]: done

**Comment by rcaneill** [comment]
> I think that the only thing left is how to deal with the units `degree_east` and `degree_north`. If the user uses cf_xarray registry it will not be a problem, but how to deal with the other case?

I guess that the only issue is for geostrophy, function that return the degree units (for inputs, if the users use pint-xarray or pint, they will have a unit registry that implement these units).
In this case, I see 4 options:
1. raise an error, saying that the registry should implement degree_east and north
2. let pint complain
3. return a non Quantity
4. return a quantity with a new registry
5. 
I think that 2. and 4. are bad options, and 3 not very good neither. So I would go for **1.**

**Comment by rcaneill** [comment]
The function gsw.geostrophic_velocity has lon and lat as arguments, which means that if the users use it with Quantities, it means that they have a registry that authorizes degree_east and degree_north.
But we still have a problem, as cf units allow both degree_east and degrees_east.

**Comment by rcaneill** [comment]
And there is the 'SK' with unit 'parts per thousand, ppt' left also...

**Comment by rcaneill** [comment]
I think that this is now ready to merge, unless you find a bug / error / something to improve

File: gsw_xarray/_core.py [hunk 3]
```diff
@@ -36,57 +38,97 @@ def quantify(rv, attrs, unit_registry=None):
...
+def pint_compat(fname, kwargs):
+    """
+    Will convert to proper unit if Quantities are used, and dequantify arguments
+
+    fname : name of the function
+    kwargs : dict of arguments and keyword arguments given to the function
+        by the user
+    """
     if pint_xarray is None:
         return args, kwargs, None
```
Comment (DocOtak) [inline gsw_xarray/_core.py:89]: args is now not used

**Comment by rcaneill** [comment]
> Wow, I really like how much cleaner the check functions are with everything as kwargs.

:)

> For the lon/lat units, internally cf-xarray [defines these in terms of degrees](https://github.com/xarray-contrib/cf-xarray/blob/33ce60bef580c8ae4e5c45f5a034e9bcb66f1a0e/cf_xarray/units.py#L98-L103) which leads to these thing being considered equivalent:
> 
> ```python
> >>> de = ureg.Quantity(1, "degrees_east")
> >>> de.to("degrees_north")
> <Quantity(1.0, 'degrees_north')>
> >>> de.to("degree")
> <Quantity(1.0, 'degree')>
> >>> de.to("radian")
> <Quantity(0.0174532925, 'radian')>
> ```
> 
> Degree north or east is the correct unit for CF, but pint really seems to only be concerned about dimensionally. The same kinda applies to the Knudsen Salinity, it's really a "unitless mass fraction", so we might just want to call that "1" internally. CF used to have the units of practical salinity be 0.001... I don't quite recall the debate that changed it to "1" though...

Ok, so to summarize: we try to use exact unit (e.g. 'degree_east', 'degree_north', 'ppt') and if they are not in the registry we fall back to either 'degree' or '1'.
 
> Is there a good place in the code that the presence/absence of the "more correct" units in the unit registry could be checked for? Falling back to something "less correct" or more generic. (We can do this in a followup PR)

I think that I can implement it

> Is geostrophic_velocity the only function that has lon/lat as outputs?

yes
 
> I think we can merge after that one args line is fixed. Any ideas on how we can test the presence/absence of optional dependencies?

I don't understand your question about dependencies, what do you mean?

**Comment by DocOtak** [comment]
My question about dependencies is wondering how we can test the situation where pint or pint-xarray is not installed. Do we emulate this with something like monkey patching? Or maybe just set up a "minimal dependencies" CI test env?

We don't need to sort this one out in this PR.

**Comment by rcaneill** [comment]
Okay I see. No idea...

**Comment by rcaneill** [comment]
I found some more bugs in my work + some problems in gsw, let me fix those and I hope that it is soon ready to merge

**Comment by DocOtak** [comment]
Indeed, I saw your open issues in the upstream gsw-python library. We might just need to convert everything back into args before calling the upstream gsw functions for compatibility reasons with most peoples existing environments.

**Comment by DocOtak** [comment]
Also, "degree" as a unit exists in the base pint unit registry so we can rely on it existing I think.

**Comment by rcaneill** [comment]
I will open an issue in gsw for this

**Comment by rcaneill** [comment]
So I propose that we stick to using kwargs, but we transform iterable into arrays internally

**Review by DocOtak — APPROVED** [review]

**Comment by DocOtak** [comment]
I think it's ready to go in... anything else you can think of?

**Pull request merged**